composer/datasets/c4.py

```diff
@@ -1,29 +1,111 @@
 # Copyright 2022 MosaicML Composer authors
 # SPDX-License-Identifier: Apache-2.0
 
-"""C4 (Colossal Cleaned CommonCrawl) dataset.
+"""C4 (Colossal Cleaned Common Crawl) dataset.
 
 This dataset is a colossal, cleaned version of Common Crawl's web crawl corpus and it is based on the `Common Crawl
 <https://commoncrawl.org>`_ dataset.
 """
 import copy
 import logging
+import os
 from functools import partial
 from itertools import chain, cycle
+from typing import Any, Dict, Optional
 
 from torch.utils.data import IterableDataset, get_worker_info
 
+from composer.datasets.streaming import StreamingDataset
 from composer.utils import dist
 from composer.utils.import_helpers import MissingConditionalImportError
 
 log = logging.getLogger(__name__)
 
-__all__ = ["C4Dataset"]
+__all__ = ["C4Dataset", "StreamingC4"]
+
+
+class StreamingC4(StreamingDataset):
+    """
+    Implementation of the C4 (Colossal Cleaned Common Crawl) dataset using StreamingDataset.
+
+    Args:
+        remote (str): Remote directory (S3 or local filesystem) where dataset is stored.
+        local (str): Local filesystem directory where dataset is cached during operation.
+        split (str): The dataset split to use, either 'train' or 'val'.
+        shuffle (bool): Whether to shuffle the samples in this dataset.
+        tokenizer_name (str): The name of the HuggingFace tokenizer to use to tokenize samples.
+        max_seq_len (int): The max sequence length of each token sample.
+        group_method (str): How to group text samples into token samples. Currently only supporting ``'truncate'``.
+        batch_size (Optional[int]): Hint batch_size that will be used on each device's DataLoader. Default: ``None``.
+    """
+
+    def _decode(self, data: bytes) -> str:
+        return data.decode('utf-8')
+
+    def _tokenize(self, text_sample):
+        if self.group_method == "truncate":
+            truncation = True
+            padding = 'max_length'
+            max_length = self.max_seq_len
+        else:
+            truncation = False
+            padding = False
+            max_length = None
+        return self.tokenizer(text_sample["text"], truncation=truncation, padding=padding, max_length=max_length)
+
+    def __init__(self,
+                 remote: str,
+                 local: str,
+                 split: str,
```

Comment:
Yeah so I want to avoid any notion of filepath structure in the base class, but then `StreamingC4` is meant to be a user-defined class that has user-specific flags, like split. I can also imagine someone building a dataset object that has no splits, or has `partition: int` or something like that, so I'd like to leave that flexibility to them.